rascaline/src/calculators/lode/spherical_expansion.rs

```diff
@@ -327,24 +394,207 @@ impl CalculatorBase for LodeSphericalExpansion {
+            let species = system.species()?;
             let cell = system.cell()?;
             if cell.shape() == UnitCell::infinite().shape() {
                 return Err(Error::InvalidParameter("LODE can only be used with periodic systems".into()));
             }
 
             let k_vectors = compute_k_vectors(&cell, k_cutoff);
+            if k_vectors.is_empty() {
+                return Err(Error::InvalidParameter("No k-vectors for current combination of hyper parameters.".into()));
+            }
+
             assert!(!k_vectors.is_empty());
 
-            self.project_k_to_nlm(&k_vectors, do_gradients);
+            self.project_k_to_nlm(&k_vectors);
+
+            let structure_factors = compute_structure_factors(
+                system.positions()?,
+                system.species()?,
+                &k_vectors
+            );
+
+            let density_fourrier = self.compute_density_fourrier(&k_vectors);
 
-            let structure_factors = compute_structure_factors(system.positions()?, &k_vectors);
+            let global_factor = 4.0 * std::f64::consts::PI / cell.volume();
+
+            for spherical_harmonics_l in 0..=self.parameters.max_angular {
+                let phase = if spherical_harmonics_l % 2 == 0 {
+                    (-1.0_f64).powi(spherical_harmonics_l as i32 / 2)
+                } else {
+                    (-1.0_f64).powi((spherical_harmonics_l as i32 + 1) / 2)
+                };
+
+                let sf_per_center = if spherical_harmonics_l % 2 == 0 {
+                    &structure_factors.real_per_center
+                } else {
+                    &structure_factors.imag_per_center
+                };
+
+                let k_vector_to_m_n = &self.k_vector_to_m_n[spherical_harmonics_l];
+
+                for (&species_neighbor, sf_per_center) in sf_per_center.iter() {
+                    for center_i in 0..system.size()? {
+                        let block_i = descriptor.keys().position(&[
+                            spherical_harmonics_l.into(),
+                            species[center_i].into(),
+                            species_neighbor.into(),
+                        ]).expect("missing block");
+
+                        let mut block = descriptor.block_mut(block_i);
+                        let values = block.values_mut();
+                        let array = values.data.as_array_mut();
+
+                        let sample = [system_i.into(), center_i.into()];
+                        let sample_i = match values.samples.position(&sample) {
+                            Some(s) => s,
+                            None => continue
+                        };
+
+                        // Add k = 0 for (m, l) = (0, 0)
+                        // For Gaussian potentials (potential_exponent = 0), the Fourier transform at k = 0 is
+                        // finite and contributes to the coefficients
+                        if self.parameters.potential_exponent == 1 {
```

Comment:
this should be potential_exponent == 0
```suggestion
                        // Add k = 0 contributions for (m, l) = (0, 0)
                        if self.parameters.potential_exponent == 0 {
                            // For Gaussian potentials (potential_exponent = 0), the Fourier transform at k = 0 is
                            // finite and contributes to the coefficients
```